### PR TITLE
Add search to docs

### LINKED ISSUE #782: Search content in the documentation
#### Summary
Search function is common and essential function for any documentation.
...

#### Why do we need this?
The documentation must be searchable to help finding document related to a topic and find problem solving solution faster.

...

#### What is already there? What do you see now?
Nothing
...

#### What is missing? What do you want to see?
A search bar proposing result based on the input
...

#### Environment
Hugo, HTML, JS
...

#### How do you propose to implement this?
Hugo propose several option on there website.
https://gohugo.io/tools/search/
Algolia looks like a good choice for the doc hosted on github, they a free Community tiers that can fit.
A local search for offline documentation would be interesting too.
...

#### Can you do this yourself and submit a Pull Request?
Yes
...

### PR DESCRIPTION
**Pull request by benolayinka**
#### Summary

Closes #782 

#### Changes

- Add search to docs
- Add gh-pages config so `mage docs:build` uses separate envs

Screenshot 2020-06-09 at 12 01 42
Screenshot 2020-06-09 at 12 01 36

#### Testing

Run locally

##### Regressions

...

#### Notes for Reviewers

Beware when testing - the results return links to the thethingsstack.io, so clicking on a result will take you out of your local environment. This may also be an issue in the future with versioning, but Algolia crawls every 24hrs so hopefully it updates ok.

DocSearch is free forever from Algolia. They crawl automatically, and create a hosted index, with some js and an API key to access it. The only requirement is that we leave their logo in the search bar.

Todo:
- What's the right place for the stylesheet and scripts? @htdvisser 
- We are already out of space on the top nav bar, what should we do? @pierrephz 
- Config.toml to enable only for some deployments @htdvisser 
- Manage separate index for TTES @benolayinka 

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md`. The target branch is set to `master` if the changes are fully compatible with existing API, database, configuration and CLI.
- [x] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [ ] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Comment by benolayinka** [comment]
Looks like I didn't rebase correctly, will fix.

@pierrephz this is the out of space issue #2415

File: doc/themes/the-things-stack/layouts/partials/head.html [hunk 1]
```diff
@@ -8,6 +8,7 @@
   <link rel="icon" type="image/png" href="{{ "favicon.png" | relURL }}">
   {{- $style := resources.Get "css/theme.scss" | resources.ToCSS (dict "includePaths" (slice "node_modules")) | resources.Minify | resources.Fingerprint }}
   <link rel="stylesheet" href="{{ $style.Permalink }}" integrity="{{ $style.Data.Integrity }}">
+  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.css" />
```
Comment (htdvisser) [inline doc/themes/the-things-stack/layouts/partials/head.html:11]: We need to make sure our Content Security Policy allows this domain for CSS.

File: doc/themes/the-things-stack/layouts/partials/nav.html [hunk 2]
```diff
@@ -39,3 +42,12 @@
     </div>
   </div>
 </nav>
+
+<script type="text/javascript" src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js"></script>
```
Comment (htdvisser) [inline doc/themes/the-things-stack/layouts/partials/nav.html:46]: We need to make sure our Content Security Policy allows this domain for JS.

File: doc/themes/the-things-stack/layouts/partials/nav.html [hunk 3]
```diff
@@ -39,3 +42,12 @@
...
+<script type="text/javascript"> docsearch({
+apiKey: '4372282a02970d8402d43457bf95d486',
```
Comment (htdvisser) [inline doc/themes/the-things-stack/layouts/partials/nav.html:48]: Can we please not hardcode API keys?

File: doc/themes/the-things-stack/layouts/shortcodes/ttnv2.html [hunk 4]
```diff
@@ -0,0 +1 @@
+The Things Network Stack V2
```
Comment (htdvisser) [inline doc/themes/the-things-stack/layouts/shortcodes/ttnv2.html:1]: This seems unrelated to the rest of the pull request.

File: doc/themes/the-things-stack/layouts/partials/nav.html (hunk 3 above)
Comment (benolayinka) [reply doc/themes/the-things-stack/layouts/partials/nav.html:48]: added to params in config.toml, but I don't know what the right way is to include it from the environment. we also want to be able to change the index for tts vs ttes

File: doc/themes/the-things-stack/layouts/partials/nav.html (hunk 2 above)
Comment (benolayinka) [reply doc/themes/the-things-stack/layouts/partials/nav.html:46]: where is the CSP?

**Comment by pierrephz** [comment]
Regarding the search bar display,
<br/>
This is an important feature for documentation, and since we already have a full navbar, I suggest to give it a full dedicated section just below the navbar, (not fixed)
<br/>
See here a mockup of what I mean @benolayinka:
<br/>
Screenshot 2020-05-26 at 13 13 30

File: doc/config.toml [hunk 5]
```diff
@@ -28,6 +28,10 @@ pygmentsUseClasses = true
   github_repository = "https://github.com/TheThingsNetwork/lorawan-stack"
   github_repository_edit = "https://github.com/TheThingsNetwork/lorawan-stack/edit/master/doc/content"
   version = "v3.8.1"
+  [params.search]
+    enabled = true
+    api_key = "4372282a02970d8402d43457bf95d486"
```
Comment (kschiffer) [inline doc/config.toml:33]: You'll need to source this from an environment variable. See: https://gohugo.io/getting-started/configuration/#configure-with-environment-variables

API keys should never be committed since now anyone can read it and use it for themselves. So in addition to sourcing this from ENV, you should generate a new API key and invalidate this one.

File: doc/config.toml.tmpl [hunk 6]
```diff
@@ -25,6 +25,10 @@ pygmentsUseClasses = true
   github_repository = "https://github.com/TheThingsNetwork/lorawan-stack"
   github_repository_edit = "https://github.com/TheThingsNetwork/lorawan-stack/edit/master/doc/content"
   version = "{{.CurrentVersion}}"
+  [params.search]
+    enabled = true
+    api_key = "4372282a02970d8402d43457bf95d486"
```
Comment (kschiffer) [inline doc/config.toml.tmpl:30]: Idem.

File: doc/config.toml (hunk 5 above)
Comment (benolayinka) [reply doc/config.toml:33]: But how do we manage our environmental variables currently? Would this be the only one?

Comment (benolayinka) [reply doc/config.toml:33]: According to Algolia, this API key is search only and is safe to publish. They won't issue a new key.

**Comment by benolayinka** [comment]
made the hugo params environment variables. @htdvisser i need to add them to travis for build, will sync with you for that.

@kschiffer how do we use material design icons? i included an mdi icon in the search bar using a css class but it looks like we don't load the resources for it.

localhost_1313_(Laptop with HiDPI screen) (3)

localhost_1313_(Laptop with HiDPI screen) (4)

**Comment by pierrephz** [comment]
Regarding the design, they are a few things that need to be changed:

### 1 - Background

Keep the same size for the container, but change the image background to this one:

bg_doc

I suggest you to use this css style:
```
.search_container{
  height: 150px;
  background-image: url(bg_doc.png);
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;
  display: flex;
  align-items: center;
  justify-content: center;
}
```

### 2 - Search bar

Make it more impactful

```
.search_bar{
  background-color: white;
  width: 55vw;
  padding: 20px;
  border-radius: 5px;
  filter: drop-shadow(0 2px 10px rgba(0, 0, 0, 0.2));
  font-size: 16px;
  color: #c7c7c7;
}
```

I did some tests and if you use those two code snippets that should be fine

image

**Comment by benolayinka** [comment]
> Without changes to `.travis.yml` this will also include search in the embedded documentation of our releases, which may get deployed to `localhost` or self-hosted deployments. I don't think that's what we want.

The HUGO_PARAMS_SEARCH_ENABLED variable disables search by default

**Comment by htdvisser** [comment]
But we just added it to our Travis configuration, which means that it will be set for both [the "internal" build](https://github.com/TheThingsNetwork/lorawan-stack/blob/develop/tools/mage/docs.go#L72-L74) and [the build for GitHub Pages](https://github.com/TheThingsNetwork/lorawan-stack/blob/develop/tools/mage/docs.go#L92).

**Comment by benolayinka** [comment]
> But we just added it to our Travis configuration, which means that it will be set for both [the "internal" build](https://github.com/TheThingsNetwork/lorawan-stack/blob/develop/tools/mage/docs.go#L72-L74) and [the build for GitHub Pages](https://github.com/TheThingsNetwork/lorawan-stack/blob/develop/tools/mage/docs.go#L92).

Can we specify an environment just for GH pages?

**Comment by htdvisser** [comment]
In Travis that isn't so easy, but @rvolosatovs is working on migrating to GitHub Actions, which makes that much easier.

**Comment by benolayinka** [comment]
@pierrephz we don't have an icon library and when I included one it didn't size properly with your custom input width so I removed it. How is this?

Screenshot 2020-06-08 at 13 30 15

**Comment by pierrephz** [comment]
@benolayinka, looks better

So if you cannot size the search bar, just decrease the height of the container and it should be fine

`height: 120px;` instead of `height: 150px;` I think

Also, from your screenshots, we can see that the suggestion bubble/popup from the search bar is aligned to the right. Shouldn’t it be aligned to the left?

**Comment by benolayinka** [comment]
@pierrephz 

Screenshot 2020-06-09 at 12 01 36

Screenshot 2020-06-09 at 12 01 42

**Comment by laurensslats** [comment]
Printscreens look good, I'm not able to run this locally though

**Comment by pierrephz** [comment]
> I don't know how happy users would be if we shove a full megabyte of `doc/themes/the-things-stack/static/img/search_bg.png` down their throats. Can't we get an SVG?

sure

@benolayinka, here is the image


[bg_doc.svg.zip](https://github.com/TheThingsNetwork/lorawan-stack/files/4752797/bg_doc.svg.zip)

**Comment by kschiffer** [comment]
> [**@kschiffer**](https://github.com/kschiffer) how do we use material design icons? i included an mdi icon in the search bar using a css class but it looks like we don't load the resources for it.

We currently don't source those in the docs. If you want to use them you need to apply

```diff
diff --git a/doc/themes/the-things-stack/assets/css/theme.scss b/doc/themes/the-things-stack/assets/css/theme.scss
index cc6f87359..c68a49c59 100644
--- a/doc/themes/the-things-stack/assets/css/theme.scss
+++ b/doc/themes/the-things-stack/assets/css/theme.scss
@@ -22,6 +22,14 @@
   font-style: normal;
 }
 
+@font-face {
+  font-family: 'Material Icons'
+  font-style: normal
+  font-weight: 400
+  font-display: block
+  src: url(../fonts/materialicons.woff2) format('woff2')
+}
+
 @import "bulma/sass/utilities/initial-variables.sass";
 
 $family-sans-serif: "Lato", "LatoFallback", $family-sans-serif;
@@ -168,6 +176,23 @@ $table-cell-padding: 1.2em 0.75em;
   }
 }
 
+.material-icon {
+  font-family: 'Material Icons';
+  font-weight: normal;
+  font-style: normal;
+  font-size: 1.3rem;
+  line-height: 1;
+  letter-spacing: normal;
+  text-transform: none;
+  display: inline-block;
+  white-space: nowrap;
+  word-wrap: normal;
+  direction: ltr;
+  vertical-align: middle;
+  font-feature-settings: 'liga';
+  align-items: center;
+}
+
 .hero.is-primary {
   background-image: url(../img/hero-bg.svg);
   background-size: cover;
```

And copy the font from `/pkg/webui/assets/fonts` to `/doc/themes/static/fonts`.

The icons can then be used with the `.material-icons` class.

File: doc/themes/the-things-stack/assets/css/theme.scss [hunk 7]
```diff
@@ -248,4 +248,25 @@ details + blockquote {
...
+}
+
+.search-bar {
+  background-color: white;
+  width: 55vw;
+  padding: 20px;
+  border-radius: 5px;
+  filter: drop-shadow(0 2px 10px rgba(0, 0, 0, 0.2));
+  font-size: 16px;
+  left: -1px; //hack for left-justified dropdown
```
Comment (kschiffer) [inline doc/themes/the-things-stack/assets/css/theme.scss:269]:
```suggestion
  left: -1px; // Hack for left-justified dropdown.
```

https://github.com/TheThingsNetwork/lorawan-stack/blob/develop/DEVELOPMENT.md#comments

File: doc/themes/the-things-stack/layouts/partials/search.html [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+      </div>
+    </div>
+</section>
+
+<script type="text/javascript" src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js"></script>
+<script type="text/javascript"> docsearch({
+apiKey: {{ .Site.Params.search.apikey }},
+indexName: {{ .Site.Params.search.index }},
+inputSelector: '#search-input',
+debug: false // Set debug to true if you want to inspect the dropdown
```
Comment (kschiffer) [inline doc/themes/the-things-stack/layouts/partials/search.html:21]:
```suggestion
debug: false // Set debug to true if you want to inspect the dropdown.
```

Idem.

**Comment by benolayinka** [comment]
> My previous comment still needs to be addressed. We need to make sure we only set the search for online (github-pages) docs, and not for the offline documentation build. This either needs to be fixed in Travis or we need to wait for GitHub Actions.

Yep blocked on this

**Comment by benolayinka** [comment]
> My previous comment still needs to be addressed. We need to make sure we only set the search for online (github-pages) docs, and not for the offline documentation build. This either needs to be fixed in Travis or we need to wait for GitHub Actions.

@htdvisser actually if we push the pre built docs with search functionality, what does it hurt? in general i'm for fixing this but if it's going to be long and painful, maybe we don't need to block this pr if the consequence is minimal

**Comment by benolayinka** [comment]
> My previous comment still needs to be addressed. We need to make sure we only set the search for online (github-pages) docs, and not for the offline documentation build. This either needs to be fixed in Travis or we need to wait for GitHub Actions.

Added a gh-pages only `config.toml` so the pages build is
`hugo ... --config gh-pages.toml,config.toml`

https://gohugo.io/getting-started/configuration/#configuration-file
